draft-ietf-quic-transport.md

```diff
@@ -1482,78 +1508,83 @@ A stream in the "open" state may be used by both peers to send frames of any
-(local)".  An endpoint receiving a FIN flag causes the stream state to become
-"half-closed (remote)"; the receiving endpoint MUST NOT process the FIN flag
-until all preceding data on the stream has been received.
-
-Either endpoint can send a RST_STREAM frame from this state, causing it to
-transition immediately to "closed".
+From this state, either endpoint can send a STREAM frame with the FIN flag set
+or a RST_STREAM frame.  These indicate the clean or abrupt termination of data
+flow on the stream, respectively.  In either case, this causes the stream to
+transition into one of the "half-closed" states.  An endpoint sending an FIN
+flag or a RST_STREAM frame causes the stream state to become "half-closed
+(local)".  An endpoint receiving a FIN flag or a RST_STREAM frame causes the
+stream state to become "half-closed (remote)".  The receiving endpoint MUST NOT
+process the FIN flag until all preceding data on the stream has been received,
+but MAY process a RST_STREAM frame immediately and discard any preceding data.
+
+Either endpoint can send a REQUEST_RST frame in this state to request closure of
+the stream in the opposite direction.  This typically indicates that the
+receiving application is no longer reading from the stream and all future data
+will be discarded upon receipt.  An endpoint SHOULD respond to a REQUEST_RST
+with a RST_STREAM of type QUIC_RECEIVED_RST.
 
 ### half-closed (local)
 
 A stream that is in the "half-closed (local)" state MUST NOT be used for sending
-STREAM frames; WINDOW_UPDATE and RST_STREAM MAY be sent in this state.
+STREAM frames; WINDOW_UPDATE, RST_STREAM, and REQUEST_RST MAY be sent in this
+state.
 
-A stream transitions from this state to "closed" when a frame that contains an
-FIN flag is received or when either peer sends a RST_STREAM frame.
+A stream transitions from this state to "closed" when a STREAM frame that
+contains an FIN flag or when a RST_STREAM frame is received.
 
 An endpoint can receive any type of frame in this state.  Providing flow-control
 credit using WINDOW_UPDATE frames is necessary to continue receiving
 flow-controlled frames.  In this state, a receiver MAY ignore WINDOW_UPDATE
 frames for this stream, which might arrive for a short period after a frame
 bearing the FIN flag is sent.
 
+If an endpoint is no longer interested in the data being received in this state,
+it MAY send a REQUEST_RST frame to request that the peer close the stream.
+STREAM frames received after sending REQUEST_RST are still counted toward the
+connection and stream flow-control windows.  Even though these frames might be
+ignored, because they are sent before their sender receives the REQUEST_RST, the
+sender will consider the frames to count against its flow-control windows.
+
+If a REQUEST_RST frame is received on a stream that is already in the
+"half-closed (local)" state, a RST_STREAM frame SHOULD still be sent,
+so that no STREAM frames previously sent will be retransmitted.
```

Comment:
I'm not sure what would be clearer.  The spec does say:

> When an endpoint sends a RST_STREAM frame, data outstanding on that stream SHOULD NOT be retransmitted, since subsequent data on this stream is expected to not be delivered by the receiver.

Maybe just "and"?